ledger/shelley/shelley.go

```diff
@@ -555,16 +566,47 @@ type ShelleyTransaction struct {
 }
 
 func (t *ShelleyTransaction) UnmarshalCBOR(cborData []byte) error {
-	type tShelleyTransaction ShelleyTransaction
-	var tmp tShelleyTransaction
-	if _, err := cbor.Decode(cborData, &tmp); err != nil {
+	// Decode as raw array to preserve metadata bytes
+	var txArray []cbor.RawMessage
+	if _, err := cbor.Decode(cborData, &txArray); err != nil {
 		return err
 	}
-	*t = ShelleyTransaction(tmp)
+
+	// Ensure we have at least 2 components
+	if len(txArray) < 2 {
+		return fmt.Errorf(
+			"invalid transaction: expected at least 2 components, got %d",
+			len(txArray),
+		)
+	}
+
+	// Decode body
+	if _, err := cbor.Decode([]byte(txArray[0]), &t.Body); err != nil {
+		return fmt.Errorf("failed to decode transaction body: %w", err)
+	}
+
+	// Decode witness set
+	if _, err := cbor.Decode([]byte(txArray[1]), &t.WitnessSet); err != nil {
+		return fmt.Errorf("failed to decode transaction witness set: %w", err)
+	}
+
+	// Handle metadata (component 3)
+	// DecodeAuxiliaryDataToMetadata already preserves raw bytes via DecodeMetadatumRaw
+	if len(txArray) > 2 && len(txArray[2]) > 0 {
```

Comment:
The TX length is not variable. The last element will always exist, and will either contain metadata or a, CBOR `nil`